src/utils/FileEmitter.ts

```diff
@@ -28,33 +28,35 @@ import { extname } from './path';
 import { isPathFragment } from './relativeId';
 import { makeUnique, renderNamePattern } from './renderNamePattern';
 
-function generateAssetFileName(
+function getSourceHash(source: string | Uint8Array, size?: number): string {
+	return createHash()
+		.update(source)
+		.digest('hex')
+		.slice(0, Math.max(0, size || defaultHashSize));
+}
+
+function pregenerateAssetFileName(
 	name: string | undefined,
 	source: string | Uint8Array,
-	outputOptions: NormalizedOutputOptions,
-	bundle: OutputBundleWithPlaceholders
-): string {
+	outputOptions: NormalizedOutputOptions
+): { fileName: string; sourceHash: string } {
+	let sourceHash: string | undefined;
 	const emittedName = outputOptions.sanitizeFileName(name || 'asset');
-	return makeUnique(
-		renderNamePattern(
-			typeof outputOptions.assetFileNames === 'function'
-				? outputOptions.assetFileNames({ name, source, type: 'asset' })
-				: outputOptions.assetFileNames,
-			'output.assetFileNames',
-			{
-				ext: () => extname(emittedName).slice(1),
-				extname: () => extname(emittedName),
-				hash: size =>
-					createHash()
-						.update(source)
-						.digest('hex')
-						.slice(0, Math.max(0, size || defaultHashSize)),
-				name: () =>
-					emittedName.slice(0, Math.max(0, emittedName.length - extname(emittedName).length))
-			}
-		),
-		bundle
+	const fileName = renderNamePattern(
+		typeof outputOptions.assetFileNames === 'function'
+			? outputOptions.assetFileNames({ name, source, type: 'asset' })
+			: outputOptions.assetFileNames,
+		'output.assetFileNames',
+		{
+			ext: () => extname(emittedName).slice(1),
+			extname: () => extname(emittedName),
+			hash: size => (sourceHash = getSourceHash(source, size)),
+			name: () =>
+				emittedName.slice(0, Math.max(0, emittedName.length - extname(emittedName).length))
+		}
 	);
+	sourceHash ??= getSourceHash(source);
```

Comment:
I thought: Why not just `const sourceHash = getSourceHash(source)` at the top? But then I noticed that we allow specifying a hash size. Now this raises a problem, though an extremely edge case one: If for some reason with a dynamic asset file name function two assets with the same source are given file names with different hash length, they are not recognized as identical.
Going even deeper, why would we even be calling the `assetFileNames` function twice in such a situation? What we actually want is that we first calculate the hash, then compare the hash, and only if it does not match we call the function for the name. For this to work reliably, the first hash needs to have the maximum hash length (e.g. by passing `Infinity` as the size, which means we do not slice). What do you think?

Reply:
Yes, the contortions in the code were because of trying to reuse the dynamic hash size.

Your idea is great, implemented it in [refactor: precompute hash then slice](https://github.com/rollup/rollup/pull/4712/commits/7e41f02915eb2a7a7b293947a325a05594d8188a)

Reply:
A lot easier to directly check the final changelog, as the code is now a lot closer to the original https://github.com/rollup/rollup/pull/4712/files